### ISSUE TITLE
Security Key Yubikey - Not working

### ISSUE THREAD

**Issue by Tomatoschewps** [Ylianst/MeshCentral#1641]
Hi,
I've tried 2FA and yubikey just now.
2FA is ok on Chrome and Firefox.

But for yubikey : 

- I can't add a yubikey from firefox. I click to "manage key", "add a key" I named it, enter my pin, touch the key and after that I have an error message.
- I can add a yubikey from chrome but once the key is added, on logging it ask for a connection token but not for the yubikey. If I have configured connection token I can logging with it, if not I can't logging at all. So the yubikey is completly ignored (chrome et firefox logging)

**Comment by Ylianst** [comment]
Ha yes, I have known about this for a while. Any FIDO2 key with a PIN does not seem to work. I need to fix that. I have one of these myself. This is not specific to Yubikey.

**Comment by Tomatoschewps** [comment]
Ok, I will check if we can use FIDO2 without pin in the meantime.
________________________________
De : Ylian Saint-Hilaire <notifications@github.com>
Envoyé : vendredi 24 juillet 2020 19:40
À : Ylianst/MeshCentral
Cc: Damien Brochard; Author
Objet : Re: [Ylianst/MeshCentral] Security Key Yubikey - Not working (#1641)


Ha yes, I have known about this for a while. Any FIDO2 key with a PIN does not seem to work. I need to fix that. I have one of these myself. This is not specific to Yubikey.

**Comment by Ylianst** [comment]
Yes, I got piles of older keys from a bunch of different vendors and they should all work fine.

**Comment by Ylianst** [comment]
Ok, I just fixed it. My YubiKey with PIN now registers correctly with MeshCentral. The fix will be in MeshCentral v0.5.94 when I next publish it.

![image](https://user-images.githubusercontent.com/1319013/88423028-f9a42600-cd9f-11ea-94c1-343e97738603.png)

**Comment by Ylianst** [comment]
MeshCentral v0.5.94 is published. Let me know if it works.

**Comment by Tomatoschewps** [comment]
Ok, upgrade done, thanks !
It's ok to add the yubikey from FF but if I disconnect it doesn't ask me for the key on login

**Comment by Ylianst** [comment]
If you have opted to register your browser for 30 days to skip the 2nd factor authentication, when it will not ask for it for 30 days. Try a different browser or clear the FireFox cache and try again. Let me know if that works.

**Comment by Tomatoschewps** [comment]
I've not opted to register for 30 days and I've just tested to login in Chrome with privacy mode with only security key activated for my account and it didn't ask for my key but ask for a security token again.
I've tried to reset the UDF part of my yubikey but no much luck.

**Comment by Ylianst** [comment]
Just to make sure, the key successfully registered. Then, you login out and when you log back in, you don't see the following screen at all? Or you see this screen, but don't see the "Use Security Key" button?

![image](https://user-images.githubusercontent.com/1319013/88468011-8fb87900-ce92-11ea-848b-3c23fd4e58e6.png)

**Comment by Tomatoschewps** [comment]
Yes, the key successfully register, I see the screen but not the "use security key" button :
![image](https://user-images.githubusercontent.com/10291238/88468155-b0f38680-cedf-11ea-8c19-cdf2b1813211.png)

**Comment by Ylianst** [comment]
When you are on the page with the missing button, can you go in the page source code and look for the the line that starts with `var hardwareKeyChallenge = `. It should look something like this:

```
var hardwareKeyChallenge = decodeURIComponent('%7B%22type%22%3A%22webAuthn%22%2C%22keyIds%22%3A%5B%22JhtKr8DEVR7hdb2q7XMuAoOXUY7HT%2FFckQKHvneojRbeq%2FZTw7E4q4QtYw4WRDbcg2Anx9El%2B0bIy0xh2gsUGw%3D%3D%22%5D%2C%22timeout%22%3A60000%2C%22challenge%22%3A%22YRL7AAoifaguXKcN2GUAi3hm45sD%2FMqVJBnMUgu%2B2HcTgcJ78TdTP4peyVnZBWE2LEb6Uw0YKXRziffGyYUSoA%3D%3D%22%7D');
```

Let me know what you see on this line. Thanks.

**Comment by Tomatoschewps** [comment]
Looks empty :
`var hardwareKeyChallenge = decodeURIComponent('');`

**Comment by Ylianst** [comment]
Interesting. Ok, two options here. Best would be for you to run:

node node_modules/meshcentral --showusers

Then find the user with the hardware key and send me the entire user entry privately, [my contact info is here](https://www.meshcommander.com/contact-information). You can XXX out any private information. What I will do is create an identical user on my server and see why the key is not showing up.

Also, if you can let me know the exact key model and version number, I may just buy an identical one.

Thanks,
Ylian

**Comment by Tomatoschewps** [comment]
Ok, I've sent you the output for my user.

I haven't specified it so far but it's an LDAP user.
I just test my yubikey with a local user (commented ldap auth in config.json) and it works ... So it's linked with the type of user ^^

**Comment by Ylianst** [comment]
Oh! It's a LDAP user!!! Ha, that explains it. Ok, I need to look into this

**Comment by Tomatoschewps** [comment]
Ok, thanks !

**Comment by Ylianst** [comment]
Ok. I just check in the fix for this. LDAP was the "key" here. The fix will be in MeshCentral v0.5.96 when it's next published.

**Comment by Ylianst** [comment]
Published MeshCentral v0.5.96 with this fix. Let me know if it works.

**Comment by Tomatoschewps** [comment]
Yeah !
Server updated, problem fixed.
My LDAP user can now use a yubikey.
Thanks
